File: discord-music-bot/src/main/java/net/masterzach32/swagbot/guilds/GuildSettings.java
package net.masterzach32.swagbot.guilds;

import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import com.google.gson.GsonBuilder;
import com.mashape.unirest.http.exceptions.UnirestException;
import net.masterzach32.swagbot.App;
import net.masterzach32.swagbot.music.PlaylistManager;
import net.masterzach32.swagbot.music.player.*;
import net.masterzach32.swagbot.utils.Constants;
import net.masterzach32.swagbot.utils.exceptions.FFMPEGException;
import net.masterzach32.swagbot.utils.exceptions.NotStreamableException;
import net.masterzach32.swagbot.utils.exceptions.YouTubeAPIException;
import net.masterzach32.swagbot.utils.exceptions.YouTubeDLException;
import sx.blah.discord.handle.impl.events.StatusChangeEvent;
import sx.blah.discord.handle.obj.IGuild;
import sx.blah.discord.handle.obj.IUser;
import sx.blah.discord.util.DiscordException;
import sx.blah.discord.util.MissingPermissionsException;
import sx.blah.discord.util.RequestBuffer;
import sx.blah.discord.util.audio.AudioPlayer;

import javax.sound.sampled.UnsupportedAudioFileException;

public class GuildSettings {

	private transient IGuild guild;
	private transient PlaylistManager playlists;
	private transient List<String> skipIDs;
	private String guildName;
	private char commandPrefix;
	private int maxSkips, volume;
	private boolean botLocked, enableNSFWFilter, announce, changeNick;
	private String lastChannel;
    private List<String> queue;
    private StatusListener listener;

	protected GuildSettings(IGuild guild, char commandPrefix, int maxSkips, int volume, boolean botLocked, boolean nsfwfilter, boolean announce, boolean changeNick, String lastChannel, List<String> queue, StatusListener listener) {
		this.guild = guild;
		playlists = new PlaylistManager(guild.getID());
		skipIDs = new ArrayList<>();
		this.guildName = guild.getName();
		this.commandPrefix = commandPrefix;
		this.maxSkips = maxSkips;
		this.volume = volume;
		this.botLocked = botLocked;
		this.enableNSFWFilter = nsfwfilter;
		this.announce = announce;
        this.changeNick = changeNick;
        this.lastChannel = lastChannel;
        this.queue = queue;
        this.listener = listener;
	}

	public String getID() {
		return guild.getID();
	}

	public void resetSkipStats() {
		skipIDs.clear();
	}

	public void addSkipID(IUser user) {
		skipIDs.add(user.getID());
	}

	public boolean hasUserSkipped(String userID) {
		return skipIDs.contains(userID);
	}

	public int numUntilSkip() {
		return maxSkips - skipIDs.size();
	}

	public int getMaxSkips() {
		return maxSkips;
	}

	public void setMaxSkips(int maxSkips) {
		this.maxSkips = maxSkips;
	}

	public int getVolume() {
		return volume;
	}

	public void setVolume(int vol) {
		this.volume = vol;
	}

	public boolean toggleBotLocked() {
		return botLocked = !botLocked;
	}

	public boolean isBotLocked() {
		return botLocked;
	}

	public PlaylistManager getPlaylistManager() {
		return playlists;
	}

	public char getCommandPrefix() {
		return commandPrefix;
	}

	public void setCommandPrefix(char prefix) {
		this.commandPrefix = prefix;
	}

	public boolean isNSFWFilterEnabled() {
		return enableNSFWFilter;
	}

	public void toggleNSFWFilter() {
		enableNSFWFilter = !enableNSFWFilter;
	}

    public boolean shouldAnnounce() {
        return announce;
    }

    public void setShouldAnnounce(boolean announce) {
        this.announce = announce;
    }

    public boolean shouldChangeNick() {
        return changeNick;
    }

    public void setChangeNick(boolean changeNick) {
        this.changeNick = changeNick;
    }

	public String getLastChannel() {
        return lastChannel;
    }

    public void setLastChannel(String id) {
        lastChannel = id;
    }

    public void setQueue(List<String> queue) {
        this.queue = queue;
    }

    public List<String> getQueue() {
        return queue;
    }

    public StatusListener getStatusListener() {
        return listener;
    }

    public boolean dispatchStatusChangedEvent(StatusChangeEvent event) {
        return listener.passEvent(event);
    }

    public IGuild getIGuild() {
		return guild;
	}

    public GuildSettings saveSettings() {
        getPlaylistManager().save();

        List<AudioPlayer.Track> tracks = getAudioPlayer().getPlaylist();
        queue = new ArrayList<>();
        for(AudioPlayer.Track track : tracks)
            if(track != null)
                queue.add(((AudioTrack) track).getUrl());

        try {
            BufferedWriter fout = new BufferedWriter(new FileWriter(Constants.GUILD_SETTINGS + guild.getID() + "/" + Constants.GUILD_JSON));
            fout.write(new GsonBuilder().setPrettyPrinting().create().toJson(this));
            fout.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
        return this;
    }

    public GuildSettings applySettings() {
        RequestBuffer.request(() -> {
            try {
				if(!App.client.getOurUser().getDisplayName(guild).equals("SwagBot"))
                	guild.setUserNickname(App.client.getOurUser(), "SwagBot");
            } catch (MissingPermissionsException | DiscordException e) {
                e.printStackTrace();
            }
        });
        App.setVolume(App.guilds.getGuildSettings(guild).getVolume(), guild);

		try {
			if (App.client.getVoiceChannelByID(lastChannel) != null && !lastChannel.equals(""))
				App.client.getVoiceChannelByID(lastChannel).join();
		} catch (MissingPermissionsException e) {
			e.printStackTrace();
		}

		if (queue.size() > 0) {
            AudioSource source;
            for (String url : queue) {
				try {
					if (url.contains("youtube"))
						source = new YouTubeAudio(url);
					else if (url.contains("soundcloud"))
						source = new SoundCloudAudio(url);
					else
						source = new AudioStream(url);
					getAudioPlayer().queue(source.getAudioTrack(null, false));
				} catch (NotStreamableException | YouTubeAPIException | UnirestException | YouTubeDLException | IOException | FFMPEGException | UnsupportedAudioFileException e) {
					e.printStackTrace();
				}
			}
        }
        return this;
    }

    public AudioPlayer getAudioPlayer() {
        return AudioPlayer.getAudioPlayerForGuild(guild);
    }
}

File: discord-music-bot/src/main/java/net/masterzach32/swagbot/guilds/GuildSettings.kt
/*
    SwagBot - A Discord Music Bot
    Copyright (C) 2016  Zachary Kozar

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */
package net.masterzach32.swagbot.guilds

import java.io.BufferedWriter
import java.io.FileWriter
import java.io.IOException
import java.util.ArrayList

import com.google.gson.GsonBuilder
import com.mashape.unirest.http.exceptions.UnirestException
import net.masterzach32.commands4j.editMessage
import net.masterzach32.commands4j.waitAndDeleteMessage
import net.masterzach32.swagbot.App
import net.masterzach32.swagbot.music.PlaylistManager
import net.masterzach32.swagbot.music.player.*
import net.masterzach32.swagbot.utils.GUILD_JSON
import net.masterzach32.swagbot.utils.GUILD_SETTINGS
import net.masterzach32.swagbot.utils.exceptions.FFMPEGException
import net.masterzach32.swagbot.utils.exceptions.NotStreamableException
import net.masterzach32.swagbot.utils.exceptions.YouTubeAPIException
import net.masterzach32.swagbot.utils.exceptions.YouTubeDLException
import sx.blah.discord.handle.impl.events.StatusChangeEvent
import sx.blah.discord.handle.obj.IGuild
import sx.blah.discord.handle.obj.IMessage
import sx.blah.discord.handle.obj.IUser
import sx.blah.discord.util.DiscordException
import sx.blah.discord.util.MissingPermissionsException
import sx.blah.discord.util.RequestBuffer
import sx.blah.discord.util.audio.AudioPlayer

import javax.sound.sampled.UnsupportedAudioFileException

data class GuildSettings(@Transient val iGuild: IGuild, var commandPrefix: Char, var maxSkips: Int, var volume: Int, var botLocked: Boolean, var nsfwFilter: Boolean, var announce: Boolean, var changeNick: Boolean, var lastChannel: String?, var queue: MutableList<String>?, val statusListener: StatusListener) {

    @Transient val playlistManager: PlaylistManager
    @Transient private val skipIDs: MutableList<String>

    init {
        playlistManager = PlaylistManager(iGuild.id)
        skipIDs = ArrayList<String>()
    }

    fun resetSkipStats() {
        skipIDs.clear()
    }

    fun addSkipID(user: IUser) {
        skipIDs.add(user.id)
    }

    fun hasUserSkipped(userID: String): Boolean {
        return skipIDs.contains(userID)
    }

    fun numUntilSkip(): Int {
        return maxSkips - skipIDs.size
    }

    fun dispatchStatusChangedEvent(event: StatusChangeEvent): Boolean {
        return statusListener.passEvent(event)
    }

    fun saveSettings(): GuildSettings {
        playlistManager.save()

        val tracks = audioPlayer.playlist
        queue = ArrayList<String>()
        for (track in tracks)
            if (track != null)
                queue!!.add((track as AudioTrack).url)

        try {
            val fout = BufferedWriter(FileWriter("${GUILD_SETTINGS}${iGuild.id}/$GUILD_JSON"))
            fout.write(GsonBuilder().setPrettyPrinting().create().toJson(this))
            fout.close()
        } catch (e: IOException) {
            e.printStackTrace()
        }

        return this
    }

    fun applySettings(): GuildSettings {
        RequestBuffer.request {
            try {
                if (App.client.ourUser.getDisplayName(iGuild) != "SwagBot")
                    iGuild.setUserNickname(App.client.ourUser, "SwagBot")
            } catch (e: MissingPermissionsException) {
                e.printStackTrace()
            } catch (e: DiscordException) {
                e.printStackTrace()
            }
        }
        App.setVolume(App.guilds.getGuildSettings(iGuild).volume.toFloat(), iGuild)

        try {
            if (App.client.getVoiceChannelByID(lastChannel) != null && lastChannel != "")
                App.client.getVoiceChannelByID(lastChannel).join()
        } catch (e: MissingPermissionsException) {
            e.printStackTrace()
        }

        if (queue!!.size > 0) {
            var source: AudioSource
            for (url in queue!!) {
                try {
                    if (url.contains("youtube"))
                        source = YouTubeAudio(url)
                    else if (url.contains("soundcloud"))
                        continue // TODO fix soundcloud
                    else
                        source = AudioStream(url)
                    audioPlayer.queue(source.getAudioTrack(null, false))
                } catch (e: NotStreamableException) {
                    e.printStackTrace()
                } catch (e: YouTubeAPIException) {
                    e.printStackTrace()
                } catch (e: UnirestException) {
                    e.printStackTrace()
                } catch (e: YouTubeDLException) {
                    e.printStackTrace()
                } catch (e: IOException) {
                    e.printStackTrace()
                } catch (e: FFMPEGException) {
                    e.printStackTrace()
                } catch (e: UnsupportedAudioFileException) {
                    e.printStackTrace()
                }

            }
        }
        return this
    }

    val audioPlayer: AudioPlayer
        get() = AudioPlayer.getAudioPlayerForGuild(iGuild)

    @Throws(IOException::class, UnsupportedAudioFileException::class)
    fun playAudioFromAudioSource(source: AudioSource?, message: IMessage?, user: IUser) {
        if (source == null)
            return
        val player = AudioPlayer.getAudioPlayerForGuild(iGuild)
        if (message != null)
            editMessage(message, "Queuing **" + source.title + "**")
        try {
            if (source is YouTubeAudio && source.isLive) {
                if(message != null)
                    waitAndDeleteMessage(editMessage(message, user.mention() + " Could not queue **" + source.title + "**: Live Streams are currently not supported!"), 120)
                return
            } else if (source is YouTubeAudio && source.isDurationAnHour) {
                if(message != null)
                    waitAndDeleteMessage(editMessage(message, user.mention() + " Could not queue **" + source.title + "**: Video length must be less than 1 hour!"), 120)
                return
            }
            player.queue(source.getAudioTrack(user, announce))
            App.joinChannel(user, iGuild)
            if (message != null)
                waitAndDeleteMessage(editMessage(message, user.mention() + " Queued **" + source.title + "**"), 30)
            return
        } catch (e: YouTubeDLException) {
            e.printStackTrace()
            if (message != null)
                waitAndDeleteMessage(editMessage(message, user.mention() + " Could not queue **" + source.title + "**: An error occurred while downloading the video."), 120)
            return
        } catch (e: FFMPEGException) {
            e.printStackTrace()
            if (message != null)
                waitAndDeleteMessage(editMessage(message, user.mention() + " Could not queue **" + source.title + "**: An error occurred while converting to audio stream"), 120)
            return
        } catch (e: IOException) {
            e.printStackTrace()
        } catch (e: MissingPermissionsException) {
            e.printStackTrace()
        }

        if (message != null)
            waitAndDeleteMessage(editMessage(message, user.mention() + " Could not queue **" + source.title + "**: (unknown reason)"), 120)
        return
    }
}